### PR TITLE
Feature/emails template improvement

### LINKED ISSUE #1322: Email registration: Template not send translated
I've a portal portal in french and the email sent for to activate the access is in English.

### PR DESCRIPTION
**Pull request by whitx**
# Description
Allow the possibility to use translation for the emails template, also allow customizable emails template per portal profile.

# Impacts
every flow that include eamil, billing, email source, sponsor, pre-registraiton, etc.

# Issue
Fixes #1322 

# Delete branch after merge
NO

# NEWS file entries
## New Features
* Now allow emails templates per portal profile
* Now allow emails template to be added to the translation

# UPGRADE file entries

### CODE REVIEW

File: conf/locale/en/LC_MESSAGES/packetfence.po [hunk 1]
```diff
@@ -800,6 +800,189 @@ msgstr "If you believe that you must be allowed to connect then contact your loc
...
+
+msgid "If you have any questions regarding the registration process please contact your local support staff."
+msgstr ""
+
+# html/captive-portal/templates/emails/emails-guest_email_preregistration.html
+msgid "Once you have activated you will receive a username and password to use once you'll be on-site."
+msgstr ""
+
+# html/captive-portal/templates/emails/emails-guest_local_account_creation.html
+msgid "User Account"
```
Comment (cgx) [inline conf/locale/en/LC_MESSAGES/packetfence.po:906]: _User account_ already exists, please use it.

File: conf/locale/en/LC_MESSAGES/packetfence.po [hunk 2]
```diff
@@ -800,6 +800,189 @@ msgstr "If you believe that you must be allowed to connect then contact your loc
...
+msgid "Violation %s was detected on your device."
+msgstr ""
+
+msgid "OS"
+msgstr ""
+
+msgid "Hostname"
+msgstr ""
+
+msgid "MAC address"
```
Comment (cgx) [inline conf/locale/en/LC_MESSAGES/packetfence.po]: _MAC Address_ already exists, please use it.

File: conf/locale/en/LC_MESSAGES/packetfence.po (hunk 1 above)
Comment (whitx) [reply conf/locale/en/LC_MESSAGES/packetfence.po:906]: This has been addressed

**Comment by whitx** [comment]
Rebased

**Comment by extrafu** [comment]
Bump

File: lib/pf/activation.pm [hunk 3]
```diff
@@ -454,9 +455,15 @@ Send an email with the activation code
 =cut
 
 sub send_email {
-    my ($activation_code, $template, %info) = @_;
+    my ($type, $activation_code, $template, %info) = @_;
     my $logger = get_logger();
 
+    use POSIX;
```
Comment (fdurand) [inline lib/pf/activation.pm]: Probably better to define it on the top

File: lib/pf/activation.pm [hunk 4]
```diff
@@ -454,9 +455,15 @@ Send an email with the activation code
...
+    my $user_locale = clean_locale(setlocale(POSIX::LC_MESSAGES));
+    if ($type eq $SPONSOR_ACTIVATION) {
+        $logger->info('We are doing sponsor activation', $user_locale);
```
Comment (fdurand) [inline lib/pf/activation.pm]: debug

File: lib/pf/util.pm [hunk 5]
```diff
@@ -1103,20 +1105,35 @@ sub strip_username {
 sub send_email {
     my ($smtp_server, $from, $to, $subject, $template, %info) = @_;
     my $logger = get_logger();
-    my %options;
-    $options{INCLUDE_PATH} = "$conf_dir/templates/";
-    $options{ENCODING} = "utf8";
+
+    use POSIX;
```
Comment (fdurand) [inline lib/pf/util.pm]: define on top

File: lib/pf/util.pm [hunk 6]
```diff
@@ -1250,6 +1267,19 @@ sub validate_date {
...
+
+Clean the format of the locale stored
+
+=cut
+
+sub clean_locale {
+    my ($locale) = @_;
+    if( $locale =~ /^([A-Za-z_]+)\./ ) {
+        $locale = $1;
+    }
```
Comment (fdurand) [inline lib/pf/util.pm:1279]: else default locale ?

Comment (cgx) [reply lib/pf/util.pm:1279]: That's fine, we just want to get rid of any extra dot-delimited part.

Comment (fdurand) [reply lib/pf/util.pm:1279]: but if there is no dot ?

Comment (cgx) [reply lib/pf/util.pm:1279]: if there's no dot, `$locale` doesn't change :)

Comment (fdurand) [reply lib/pf/util.pm:1279]: Sure ? my $lang = clean_locale($self->app->session->{locale});

Comment (fdurand) [reply lib/pf/util.pm:1279]: i prefer to have something more explicit like:
sub clean_locale {
    my ($locale) = @_;
    if( $locale =~ /^([A-Za-z_]+)\./ ) {
        return $1;
     }
    return $locale;
}

Comment (cgx) [reply lib/pf/util.pm:1279]: Oh! Just adding `return $locale;` would fix it.

**Comment by fdurand** [comment]
Agree to merge

**Comment by lzammit** [comment]
I've tested it and it works.

I've flagged one bug #1967 related to the email registration flow and an enhancement #1966 for the phone number cleaning.

Beside that, I had to do the create the translation manually so will need to do the translation on Transfix.

**Pull request merged**